crates/aptos-logger/src/aptos_logger.rs

```diff
@@ -17,46 +17,83 @@ use aptos_infallible::RwLock;
 use once_cell::sync::Lazy;
-use serde::Serialize;
+use serde::ser::SerializeStruct;
+use serde::{Serialize, Serializer};
 use std::{
     collections::BTreeMap,
     env, fmt,
     io::Write,
+    str::FromStr,
     sync::{
         mpsc::{self, Receiver, SyncSender},
         Arc,
     },
     thread,
 };
+use strum_macros::EnumString;
 
 const RUST_LOG: &str = "RUST_LOG";
 const RUST_LOG_REMOTE: &str = "RUST_LOG_REMOTE";
+const RUST_LOG_FORMAT: &str = "RUST_LOG_FORMAT";
 /// Default size of log write channel, if the channel is full, logs will be dropped
 pub const CHANNEL_SIZE: usize = 10000;
 const NUM_SEND_RETRIES: u8 = 1;
 
+#[derive(EnumString)]
+#[strum(serialize_all = "lowercase")]
+enum LogFormat {
+    Json,
+    Text,
+}
+
 /// A single log entry emitted by a logging macro with associated metadata
-#[derive(Debug, Serialize)]
+#[derive(Debug)]
 pub struct LogEntry {
-    #[serde(flatten)]
     metadata: Metadata,
-    #[serde(skip_serializing_if = "Option::is_none")]
     thread_name: Option<String>,
     /// The program backtrace taken when the event occurred. Backtraces
     /// are only supported for errors and must be configured.
-    #[serde(skip_serializing_if = "Option::is_none")]
     backtrace: Option<String>,
-    #[serde(skip_serializing_if = "Option::is_none")]
     hostname: Option<&'static str>,
-    #[serde(skip_serializing_if = "Option::is_none")]
     namespace: Option<&'static str>,
     timestamp: String,
-    #[serde(skip_serializing_if = "BTreeMap::is_empty")]
     data: BTreeMap<Key, serde_json::Value>,
-    #[serde(skip_serializing_if = "Option::is_none")]
     message: Option<String>,
 }
 
+// implement custom serializer for LogEntry since we want to promote the `metadata.level` field into a top level field
+// and prefix the remaining metadata attributes as `source.<metadata_field>`.
+impl Serialize for LogEntry {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("LogEntry", 9)?;
+        state.serialize_field("level", &self.metadata.level())?;
+        state.serialize_field("source", &self.metadata)?;
+        if let Some(thread_name) = &self.thread_name {
+            state.serialize_field("thread_name", thread_name)?;
+        }
+        if let Some(hostname) = &self.hostname {
+            state.serialize_field("hostname", hostname)?;
+        }
+        if let Some(namespace) = &self.namespace {
+            state.serialize_field("namespace", namespace)?;
+        }
+        state.serialize_field("timestamp", &self.timestamp)?;
+        if let Some(message) = &self.message {
+            state.serialize_field("message", message)?;
+        }
+        if !&self.data.is_empty() {
+            state.serialize_field("data", &self.data)?;
+        }
+        if let Some(backtrace) = &self.backtrace {
+            state.serialize_field("backtrace", backtrace)?;
+        }
```

Comment:
I thought there was a way to specify this on the macro of the struct?

https://serde.rs/field-attrs.html#skip_serializing_if

Reply:
yes, but as the comment above says I needed to do a bit more than what was possible with macros. In particular I wanted to keep some of the Metadata fields in a nested `source` field but promote the `level` to a top-level field.
I did a bunch of research and there seems no macro-based way to achieve that.
A cleaner way would be probably to rename `Metadata` to `SourceMetadata` and put the `level` field into the LogEntry directly or something but that would require a lot more code changes than just a custom serializer.